sonarts-core/src/rules/noNestedIncdecRule.ts

```diff
@@ -0,0 +1,99 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as tslint from "tslint";
+import * as ts from "typescript";
+import { SonarRuleMetaData } from "../sonarRule";
+import { SonarRuleVisitor } from "../utils/sonarAnalysis";
+import { isBinaryExpression, isPrefixUnaryExpression, isPostfixUnaryExpression } from "../utils/nodes";
+
+export class Rule extends tslint.Rules.TypedRule {
+  public static metadata: SonarRuleMetaData = {
+    ruleName: "no-nested-incdec",
+    description:
+      "Increment (++) and decrement (--) operators should not be used in a method call or mixed with other operators in an expression",
+    rationale: tslint.Utils.dedent``,
+    optionsDescription: "",
+    options: null,
+    rspecKey: "RSPEC-881",
+    type: "functionality",
+    typescriptOnly: false,
+  };
+
+  public static MESSAGE_INC = "Extract this increment operation into a dedicated statement.";
+  public static MESSAGE_DEC = "Extract this decrement operation into a dedicated statement.";
+
+  public applyWithProgram(sourceFile: ts.SourceFile): tslint.RuleFailure[] {
+    return new Visitor(this.getOptions().ruleName).visit(sourceFile).getIssues();
+  }
+}
+
+const incDecOperators = [ts.SyntaxKind.PlusPlusToken, ts.SyntaxKind.MinusMinusToken];
+
+class Visitor extends SonarRuleVisitor {
+  public visitExpressionStatement(node: ts.ExpressionStatement) {
+    if (isIncDec(node.expression)) {
+      // skipping visit visitPrefixUnaryExpression/visitPostfixUnaryExpression if increment/decrement
+      this.visitNode((node.expression as ts.PrefixUnaryExpression).operand);
+    } else {
+      super.visitExpressionStatement(node);
+    }
+  }
+
+  public visitForStatement(node: ts.ForStatement) {
+    node.initializer && this.visitNode(node.initializer);
+    node.condition && this.visitNode(node.condition);
+    node.incrementor && this.visitIncrementor(node.incrementor);
+    this.visitNode(node.statement);
+  }
+
+  public visitPrefixUnaryExpression(node: ts.PrefixUnaryExpression) {
+    this.checkUnaryExpression(node);
+    super.visitPrefixUnaryExpression(node);
+  }
+
+  public visitPostfixUnaryExpression(node: ts.PostfixUnaryExpression) {
+    this.checkUnaryExpression(node);
+    super.visitPostfixUnaryExpression(node);
+  }
+
+  private visitIncrementor(node: ts.Expression) {
+    if (isIncDec(node)) {
+      this.visitNode((node as ts.PrefixUnaryExpression).operand);
+    } else if (isBinaryExpression(node) && node.operatorToken.kind === ts.SyntaxKind.CommaToken) {
+      const commaExpr = node as ts.BinaryExpression;
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS4325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-493&issues=AWGu5_COaHWQ5zPscma8&open=AWGu5_COaHWQ5zPscma8)

[ISSUE_KEY]: AWGu5_COaHWQ5zPscma8